File: app/services/position_match/spec.py
"""Position match core types (SlotConstraint, MatchSpec, CandidateSource)."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal, Optional, Protocol, Union


ConstraintKind = Literal[
    "code_digit",        # 0243 碼數字約束（單一 digit）
    "literal_char",      # 必須完全匹配的粵字（canto char）
    "final_anchor",      # 韻母（final）錨點（參考字的發音）
    "initial_anchor",    # 聲母（initial）錨點
    "wildcard",          # 萬用字元（_ ? %），不做約束
]


@dataclass(frozen=True)
class SlotConstraint:
    """
    單一位置的約束條件。

    這是 PositionMatchEngine 的核心資料結構，取代原本散落在多處的
    code_digits、mask、anchor、literal_char 等參數。
    """
    pos: int                                   # 該約束在詞內的位置（0-based）
    kind: ConstraintKind
    value: Optional[Union[str, set[str]]] = None   # 視 kind 而定：digit 字串、canto char、phoneme set 等


@dataclass
class MatchSpec:
    """
    一次位置型查詢的完整規格。

    由 parser 產生的資料（CodeTailQuery、RhymeAnchorQuery、MaskQuery 等）
    轉換而成，或由 handle_* 薄層建構。
    """
    width: int                                 # 目標詞長度
    slots: list[SlotConstraint] = field(default_factory=list)

    # 額外全域約束
    code_prefix: Optional[str] = None          # 例如 code_tail 的 code_digits
    literal_priority: bool = False             # mask 類查詢常用（literal 數量優先排序）
    # 預留給 hybrid 等特殊語意
    hybrid_ref_chars: Optional[str] = None
    hybrid_ref_pos: Optional[int] = None

    # 為 mask 等
    mask: str = ""

    # 等號查詢／碼夾等號查詢（CONTEXT § 等號查詢、碼夾等號查詢）
    ref_literal: str = ""
    ref_start_pos: int = 0
    ref_dimension: Literal["initial", "final"] = "final"
    phoneme_anchor_only: bool = False
    whole_word_phoneme_match: bool = False

    # 未來可擴充的政策（排序、去重策略等）
    extra: dict[str, Any] = field(default_factory=dict)


class CandidateSource(Protocol):
    """
    候選來源抽象（隱藏 cache 與 DB 的差異）。

    實作會有兩種：
    - Cache 優先來源（word_cache + get_words_for_length）
    - DB 後援來源（GLOB + length_filter + apply_code_filter）

    回傳 (candidates, used_cache)：used_cache 為 True 時 candidates 是 dict list，否則是 ORM 物件 list。
    """

    def get_candidates(
        self,
        length: int,
        *,
        code: Optional[str] = None,
        mode: str = "m1",
    ) -> tuple[list[Any], bool]:
        ...


@dataclass
class MaskFamilySearchResult:
    """缺字型查詢執行結果（供 query_dispatch 包成 SearchResult）。"""

    items: list
    cache_path: Optional[str] = None

File: client/src/db/position-match/spec.ts
/**
 * Position match core types — port of app/services/position_match/spec.py (MF-1)
 */

export type ConstraintKind =
  | 'code_digit'
  | 'literal_char'
  | 'final_anchor'
  | 'initial_anchor'
  | 'rhyme_letters'
  | 'syllable_letters'
  | 'initial_letters'
  | 'wildcard';

export type SlotConstraintValue = string | ReadonlySet<string>;

export interface SlotConstraint {
  pos: number;
  kind: ConstraintKind;
  value?: SlotConstraintValue | null;
}

/** Python EqualsSpan.dimension; TS executors also accept rhyme/phone aliases */
export type EqualsSpanDimension = 'initial' | 'final';

export type EqualsDimension = EqualsSpanDimension | 'rhyme' | 'phone';

export interface EqualsSpan {
  ref_literal: string;
  start_pos: number;
  dimension: EqualsDimension;
  phoneme_anchor_only: boolean;
  whole_word: boolean;
}

export type CompoundKind = 'syn' | 'ant' | 'doubled_syllable';

export interface MatchSpec {
  width: number;
  slots?: SlotConstraint[];
  code_prefix?: string | null;
  literal_priority?: boolean;
  hybrid_ref_chars?: string | null;
  hybrid_ref_pos?: number | null;
  mask?: string;
  compound_kind?: CompoundKind | null;
  extra?: Record<string, unknown>;
}

export interface MaskFamilySearchResult {
  items: unknown[];
  cache_path?: string | null;
}

export interface CandidateSource {
  getCandidates(
    length: number,
    options?: { code?: string | null; mode?: string },
  ): [unknown[], boolean];
}

const EQUALS_SPAN_KEY = 'equals_span';

export function getEqualsSpan(spec: MatchSpec): EqualsSpan | null {
  const raw = spec.extra?.[EQUALS_SPAN_KEY];
  if (!raw || typeof raw !== 'object') {
    return null;
  }
  return raw as EqualsSpan;
}

export function attachEqualsSpan(spec: MatchSpec, span: EqualsSpan): void {
  if (!spec.extra) {
    spec.extra = {};
  }
  spec.extra[EQUALS_SPAN_KEY] = span;
}

export function createMatchSpec(width: number, fields: Omit<MatchSpec, 'width'> = {}): MatchSpec {
  return { width, slots: [], mask: '', extra: {}, ...fields };
}

/** ponytail: runnable self-check — `npx tsx client/scripts/position-match-spec-self-check.ts` */
export function positionMatchSpecSelfCheck(): void {
  const spec = createMatchSpec(2, { code_prefix: '23', mask: '?就' });
  if (spec.width !== 2 || spec.code_prefix !== '23') {
    throw new Error('positionMatchSpecSelfCheck: createMatchSpec');
  }

  const span: EqualsSpan = {
    ref_literal: '香港',
    start_pos: 0,
    dimension: 'final',
    phoneme_anchor_only: false,
    whole_word: true,
  };
  attachEqualsSpan(spec, span);
  const roundtrip = getEqualsSpan(spec);
  if (!roundtrip || roundtrip.ref_literal !== '香港' || !roundtrip.whole_word) {
    throw new Error('positionMatchSpecSelfCheck: equals_span roundtrip');
  }

  spec.slots!.push({ pos: 1, kind: 'literal_char', value: '就' });
  if (spec.slots!.length !== 1 || spec.slots![0]!.kind !== 'literal_char') {
    throw new Error('positionMatchSpecSelfCheck: slots');
  }
}
